topostats/tracing/dnatracing.py

```diff
@@ -493,69 +488,88 @@ def get_splined_traces(self):
 
-                for i in range(step_size):
-                    x_sampled = np.array([self.fitted_traces[dna_num][:, 0][j] for j in
-                                          range(i, len(self.fitted_traces[dna_num][:, 0]), step_size)])
-                    y_sampled = np.array([self.fitted_traces[dna_num][:, 1][j] for j in
-                                          range(i, len(self.fitted_traces[dna_num][:, 1]), step_size)])
-
-                    interp_f = interp.interp1d(x_sampled, y_sampled, kind='cubic', assume_sorted=False)
+                for i in range(step_size_px):
+                    x_sampled = np.array(
+                        [
+                            self.fitted_traces[dna_num][:, 0][j]
+                            for j in range(i, len(self.fitted_traces[dna_num][:, 0]), step_size_px)
+                        ]
+                    )
+                    y_sampled = np.array(
+                        [
+                            self.fitted_traces[dna_num][:, 1][j]
+                            for j in range(i, len(self.fitted_traces[dna_num][:, 1]), step_size_px)
+                        ]
+                    )
 
-                    x_new = np.linspace(start_x, end_x, interp_step)
-                    y_new = interp_f(x_new)
+                    try:
+                        tck, u = interp.splprep([x_sampled, y_sampled], s=5, per=0, quiet=1, k=3)
+                        out = interp.splev(ev_array, tck)
+                        splined_trace = np.column_stack((out[0], out[1]))
+                    except ValueError:
+                        # Value error occurs when the "trace fitting" really messes up the traces
 
-                    print(y_new)
+                        x = np.array(
+                            [
+                                self.ordered_traces[dna_num][:, 0][j]
+                                for j in range(i, len(self.ordered_traces[dna_num][:, 0]), step_size_px)
+                            ]
+                        )
+                        y = np.array(
+                            [
+                                self.ordered_traces[dna_num][:, 1][j]
+                                for j in range(i, len(self.ordered_traces[dna_num][:, 1]), step_size_px)
+                            ]
+                        )
 
-                    # tck = interp.splrep(x_sampled, y_sampled, quiet = 0)
-                    # out = interp.splev(np.linspace(start_x,end_x, nbr*step_size), tck)
-                    splined_trace = np.column_stack((x_new, y_new))
+                        try:
+                            tck, u = interp.splprep([x, y], s=5, per=0, quiet=1)
+                            out = interp.splev(np.linspace(0, 1, nbr * step_size_px), tck)
+                            splined_trace = np.column_stack((out[0], out[1]))
+                        except ValueError:  # sometimes even the ordered_traces are too bugged out so just delete these traces
```

Comment:
Hi, Neil, thanks a lot for your review! I'll have a look at your suggestions and address the ones that are quick to fix. As for spliting the splining into linear and circular, I am not sure if this is necessary, because the process for splining linear and circular molecules are almost the same, with the only difference being the parameters used when calling `interp.splprep`; I don't think this change in numbers should merit a separate method?

Reply:
Would it be a better idea for me to merge the linear and circular splining into one part, with the if statement only applied before the `interp.splprep()` instead of the entire code block?

Reply:
Also, do we have unit tests for the splining of circular molecules? If so, I should be able to modify them to be compatible with both circular and linear molecules